src/platform/OpenThread/GenericThreadBorderRouterDelegate.cpp

```diff
@@ -0,0 +1,193 @@
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetDataset(Thread::OperationalDataset & dataset, DatasetType type)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+
+    ScopedThreadLock threadLock;
+    otError otErr = OT_ERROR_NONE;
+    otOperationalDatasetTlvs datasetTlvs;
+    if (type == DatasetType::kActive)
+    {
+        otErr = otDatasetGetActiveTlvs(otInst, &datasetTlvs);
+    }
+    else
+    {
+        otErr = otDatasetGetPendingTlvs(otInst, &datasetTlvs);
+    }
+    if (otErr == OT_ERROR_NONE)
+    {
+        return dataset.Init(ByteSpan(datasetTlvs.mTlvs, datasetTlvs.mLength));
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::SetActiveDataset(const Thread::OperationalDataset & activeDataset,
+                                                                   uint32_t sequenceNum, ActivateDatasetCallback * callback)
+{
+    // This function will never be invoked when there is an Active Dataset already configured.
+    CHIP_ERROR err = SaveThreadBorderRouterCommissioned(false);
+    if (err == CHIP_NO_ERROR)
+    {
+        err = DeviceLayer::ThreadStackMgrImpl().AttachToThreadNetwork(activeDataset, nullptr);
+    }
+    if (err == CHIP_NO_ERROR)
+    {
+        // We can change the stored sequence number because OnActivateDatasetComplete executed later must be called for this
+        // SetActiveDataset request
+        mSequenceNum = sequenceNum;
+        mCallback    = callback;
+    }
+    return err;
+}
+
+void GenericOpenThreadBorderRouterDelegate::OnPlatformEventHandler(const DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
+{
+    GenericOpenThreadBorderRouterDelegate * delegate = reinterpret_cast<GenericOpenThreadBorderRouterDelegate *>(arg);
+    if (delegate && delegate->mCallback)
+    {
+        if (event->Type == DeviceLayer::DeviceEventType::kThreadConnectivityChange &&
+            event->ThreadConnectivityChange.Result == DeviceLayer::kConnectivity_Established)
+        {
+            delegate->mCallback->OnActivateDatasetComplete(delegate->mSequenceNum, CHIP_NO_ERROR);
+            SaveThreadBorderRouterCommissioned(true);
+            delegate->mCallback = nullptr;
+        }
+    }
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::SaveThreadBorderRouterCommissioned(bool commissioned)
+{
+    return DeviceLayer::PersistedStorage::KeyValueStoreMgr().Put(kFailsafeThreadBorderRouterCommissioned, commissioned);
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::RevertActiveDataset()
+{
+    // The FailSafe Timer is triggered and the previous command request should be handled, so reset the callback.
+    mCallback               = nullptr;
+    bool threadCommissioned = true;
+    DeviceLayer::PersistedStorage::KeyValueStoreMgr().Get(kFailsafeThreadBorderRouterCommissioned, &threadCommissioned);
```

Comment:
If it's already "commissioned", why does Revert do nothing?

Reply:
The revert active dataset is called to clear active dataset stored by Thread stack. so if it is already commissioned(ActiveDataset is set by SetActiveDataset and receives CommissioningComplete command) we will do nothing. This "commissioned" check is only for the call of delegate when initializing the delegate.